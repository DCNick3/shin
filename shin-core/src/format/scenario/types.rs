use crate::format::scenario::instruction_elements::NumberSpec;
use binrw::{BinRead, BinResult, BinWrite, Endian, VecArgs};
use derivative::Derivative;
use smallvec::SmallVec;
use std::fmt;
use std::io::{Read, Seek, Write};
use std::marker::PhantomData;

// TODO: make lists generic over the type of length
/// A list of `T` with a u8 length
#[derive(Debug)]
pub struct U8List<T>(pub Vec<T>);

/// A list of `T` with a u16 length
#[derive(Debug)]
pub struct U16List<T>(pub Vec<T>);

/// A list of `T` with a length of `L`, stored in a `SmallVec` with array `A`
pub struct SmallList<L: Into<usize> + TryFrom<usize> + 'static, A: smallvec::Array>(
    pub SmallVec<A>,
    pub PhantomData<L>,
);

pub type U8SmallList<A> = SmallList<u8, A>;
pub type U16SmallList<A> = SmallList<u16, A>;

pub const SMALL_LIST_SIZE: usize = 6;

pub type U8SmallNumberList<T = i32> = U8SmallList<[NumberSpec<T>; SMALL_LIST_SIZE]>;
pub type U16SmallNumberList<T = i32> = U16SmallList<[NumberSpec<T>; SMALL_LIST_SIZE]>;

/// Pad the contents to 4 bytes
///
/// (Used in [super::Instruction::jt])
#[derive(Derivative)]
#[derivative(Debug = "transparent")]
pub struct Pad4<T>(pub T);

impl<T: for<'a> BinRead<Args<'a> = ()> + 'static> BinRead for U8List<T> {
    type Args<'a> = ();

    fn read_options<R: Read + Seek>(reader: &mut R, endian: Endian, _: ()) -> BinResult<Self> {
        let len = u8::read_options(reader, endian, ())?;

        Ok(Self(<_>::read_options(
            reader,
            endian,
            VecArgs {
                count: len as usize,
                inner: (),
            },
        )?))
    }
}
impl<T: for<'a> BinWrite<Args<'a> = ()>> BinWrite for U8List<T> {
    type Args<'a> = ();

    fn write_options<W: Write + Seek>(
        &self,
        _writer: &mut W,
        _endian: Endian,
        _: (),
    ) -> BinResult<()> {
        todo!()
    }
}

impl<T: for<'a> BinRead<Args<'a> = ()> + 'static> BinRead for U16List<T> {
    type Args<'a> = ();

    fn read_options<R: Read + Seek>(reader: &mut R, endian: Endian, _: ()) -> BinResult<Self> {
        let len = u16::read_options(reader, endian, ())?;

        Ok(Self(<_>::read_options(
            reader,
            endian,
            VecArgs {
                count: len as usize,
                inner: (),
            },
        )?))
    }
}
impl<T: for<'a> BinWrite<Args<'a> = ()>> BinWrite for U16List<T> {
    type Args<'a> = ();

    fn write_options<W: Write + Seek>(
        &self,
        _writer: &mut W,
        _endian: Endian,
        _: (),
    ) -> BinResult<()> {
        todo!()
    }
}

impl<L: Into<usize> + TryFrom<usize>, A: smallvec::Array> fmt::Debug for SmallList<L, A>
where
    A::Item: fmt::Debug,
{
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_tuple("U8SmallList").field(&self.0).finish()
    }
}

impl<
        L: Into<usize> + TryFrom<usize> + for<'a> BinRead<Args<'a> = ()>,
        A: smallvec::Array<Item = T> + 'static,
        T: for<'a> BinRead<Args<'a> = ()>,
    > BinRead for SmallList<L, A>
{
    type Args<'a> = ();

    fn read_options<R: Read + Seek>(reader: &mut R, endian: Endian, _: ()) -> BinResult<Self> {
        let len = L::read_options(reader, endian, ())?.into();

        let mut res = SmallVec::new();
        res.reserve(len);
        for _ in 0..len {
            res.push(<_>::read_options(reader, endian, ())?);
        }

        Ok(Self(res, PhantomData {}))
    }
}

impl<
        L: Into<usize> + TryFrom<usize> + for<'a> BinWrite<Args<'a> = ()>,
        A: smallvec::Array<Item = T> + 'static,
        T: for<'a> BinWrite<Args<'a> = ()>,
    > BinWrite for SmallList<L, A>
{
    type Args<'a> = ();

    fn write_options<W: Write + Seek>(
        &self,
        _writer: &mut W,
        _endian: Endian,
        _: (),
    ) -> BinResult<()> {
        todo!()
    }
}

impl<T: for<'a> BinRead<Args<'a> = ()> + 'static> BinRead for Pad4<T> {
    type Args<'a> = ();

    fn read_options<R: Read + Seek>(
        reader: &mut R,
        endian: Endian,
        args: Self::Args<'_>,
    ) -> BinResult<Self> {
        let pos = reader.stream_position()?;
        let res = <_>::read_options(reader, endian, args)?;
        let new_pos = reader.stream_position()?;

        assert!(new_pos - pos <= 4, "Pad4: read more than 4 bytes");

        reader.seek(std::io::SeekFrom::Start(pos + 4))?;

        Ok(Self(res))
    }
}
impl<T: for<'a> BinWrite<Args<'a> = ()>> BinWrite for Pad4<T> {
    type Args<'a> = ();

    fn write_options<W: Write + Seek>(
        &self,
        _writer: &mut W,
        _endian: Endian,
        _: (),
    ) -> BinResult<()> {
        todo!()
    }
}
