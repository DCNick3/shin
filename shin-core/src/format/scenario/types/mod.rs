mod list;
mod pad4;

pub use list::{
    SmallList, U16SmallList, U16SmallNumberList, U8SmallList, U8SmallNumberList, SMALL_LIST_SIZE,
};
pub use pad4::Pad4;

pub use crate::format::text::{
    SJisString, StringArray, U16FixupString, U16String, U8FixupString, U8String, ZeroString,
};
