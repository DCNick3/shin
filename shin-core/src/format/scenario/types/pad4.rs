use std::{io, io::Seek};

use binrw::{BinRead, BinResult, BinWrite, Endian};
use derivative::Derivative;

/// Pad the contents to 4 bytes
///
/// (Used in [super::Instruction::gt])
#[derive(Derivative, PartialEq, Eq, Copy, Clone)]
#[derivative(Debug = "transparent")]
pub struct Pad4<T>(pub T);

impl<T: for<'a> BinRead<Args<'a> = ()> + 'static> BinRead for Pad4<T> {
    type Args<'a> = ();

    fn read_options<R: io::Read + io::Seek>(
        reader: &mut R,
        endian: Endian,
        args: Self::Args<'_>,
    ) -> BinResult<Self> {
        let pos = reader.stream_position()?;
        let res = <_>::read_options(reader, endian, args)?;
        let new_pos = reader.stream_position()?;

        assert!(new_pos - pos <= 4, "Pad4: read more than 4 bytes");

        // read the padding bytes
        for _ in 0..(4 - (new_pos - pos)) {
            u8::read_options(reader, endian, ())?;
        }

        Ok(Self(res))
    }
}
impl<T: for<'a> BinWrite<Args<'a> = ()>> BinWrite for Pad4<T> {
    type Args<'a> = ();

    fn write_options<W: io::Write + Seek>(
        &self,
        writer: &mut W,
        endian: Endian,
        _: (),
    ) -> BinResult<()> {
        let pos = writer.stream_position()?;
        <_>::write_options(&self.0, writer, endian, ())?;
        let new_pos = writer.stream_position()?;

        assert!(new_pos - pos <= 4, "Pad4: wrote more than 4 bytes");

        // write the padding zero bytes
        for _ in 0..(4 - (new_pos - pos)) {
            0u8.write_options(writer, endian, ())?;
        }

        Ok(())
    }
}
