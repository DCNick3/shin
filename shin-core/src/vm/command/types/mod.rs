//! Types used in commands.

mod flags;
mod id;
mod property;

pub use flags::{AudioWaitStatus, LayerCtrlFlags, MaskFlags};
pub use id::{
    LayerId, LayerIdOpt, VLayerId, VLayerIdRepr, LAYERBANKS_COUNT, LAYERS_COUNT, PLANES_COUNT,
};
pub use property::{LayerProperty, LayerPropertySmallList};

use crate::format::scenario::instruction_elements::FromNumber;
use num_derive::FromPrimitive;

#[derive(FromPrimitive, Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum LayerType {
    Null = 0,
    Tile = 1,
    Picture = 2,
    Bustup = 3,
    Animation = 4,
    Effect = 5,
    Movie = 6,
    FocusLine = 7,
    Rain = 8,
    Quiz = 9,
}

impl FromNumber for LayerType {
    fn from_number(number: i32) -> Self {
        num_traits::FromPrimitive::from_i32(number)
            .unwrap_or_else(|| panic!("LayerType::from_vm_ctx: invalid layer type: {}", number))
    }
}

#[derive(FromPrimitive, Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Hash)]
pub enum MessageboxType {
    Neutral = 0,
    WitchSpace = 1,
    Ushiromiya = 2,
    Transparent = 3,
    Novel = 4,
    NoText = 5,
}

#[derive(FromPrimitive, Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Hash)]
pub enum MessageTextLayout {
    Left = 0,
    /// I _think_ this is the same as Left
    Layout1 = 1,
    Center = 2,
    Right = 3,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Hash)]
pub struct MessageboxStyle {
    pub messagebox_type: MessageboxType,
    pub text_layout: MessageTextLayout,
}

impl Default for MessageboxStyle {
    fn default() -> Self {
        Self {
            messagebox_type: MessageboxType::Neutral,
            text_layout: MessageTextLayout::Left,
        }
    }
}

impl FromNumber for MessageboxStyle {
    fn from_number(number: i32) -> Self {
        assert!(number >= 0);
        let msgbox_type = number & 0xf;
        let text_layout = (number >> 4) & 0xf;
        Self {
            messagebox_type: num_traits::FromPrimitive::from_i32(msgbox_type).unwrap_or_else(
                || panic!("MsgInit::from: unknown messagebox type: {}", msgbox_type),
            ),
            text_layout: num_traits::FromPrimitive::from_i32(text_layout)
                .unwrap_or_else(|| panic!("MsgInit::from: unknown text layout: {}", text_layout)),
        }
    }
}

/// A volume value, in the range [0.0, 1.0].
#[derive(Debug, Copy, Clone, PartialEq)]
pub struct Volume(pub f32);

impl Default for Volume {
    fn default() -> Self {
        Self(1.0)
    }
}

impl FromNumber for Volume {
    fn from_number(number: i32) -> Self {
        Self((number as f32 / 1000.0).clamp(0.0, 1.0)) // TODO: warn if out of range
    }
}

/// Defines a pan value in the range [-1.0, 1.0], where `0.0` is the center and `-1.0` is the hard left and `1.0` is the hard right.
#[derive(Debug, Copy, Clone, PartialEq)]
pub struct Pan(pub f32);

impl Default for Pan {
    fn default() -> Self {
        Self(0.0)
    }
}

impl FromNumber for Pan {
    fn from_number(number: i32) -> Self {
        Self((number as f32 / 1000.0).clamp(-1.0, 1.0)) // TODO: warn if out of range
    }
}
