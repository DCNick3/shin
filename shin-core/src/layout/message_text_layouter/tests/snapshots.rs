//! Tests based on single layout snapshots

use insta::assert_debug_snapshot;

use super::make_snapshot;
use crate::vm::command::types::{MessageTextLayout, MessageboxType};

#[test]
fn snapshot_ep1_nanjo1() {
    // here we have rubi text that is wider than the base text. the pointers have to be adjusted
    let snapshot = make_snapshot(MessageTextLayout::Justify, MessageboxType::Ushiromiya, "南條\u{3000}輝正@r@v19/11900001.「…………また。@k@v19/11900002.…お酒を@bたしな.@<嗜@>まれましたな？」");

    assert_debug_snapshot!("ep1_nanjo1", snapshot);
}

#[test]
fn snapshot_ep1_nanjo2() {
    // here we have a pretty long message with multiple Waits and spanning multiple lines
    let snapshot = make_snapshot(MessageTextLayout::Justify, MessageboxType::Ushiromiya, "南條\u{3000}輝正@r@v19/11900003.「…………金蔵さん。@k@v19/11900004.…あんたの体が一見調子がいいのは薬が効いてるからだ。@k@v19/11900005.だが、そんな強い酒を飲み続けては薬の意味もなくなってしまう。@k@v19/11900006.…悪いことは言わん。@k@v19/11900007.酒は控えなさい」");

    assert_debug_snapshot!("ep1_nanjo2", snapshot);
}

#[test]
fn snapshot_ep1_kinzo1() {
    // here we have rubi text that is smaller than the base text
    let snapshot = make_snapshot(MessageTextLayout::Justify, MessageboxType::Ushiromiya, "右代宮\u{3000}金蔵@r@v01/11500012.「はっはっはっは…。@k@v01/11500013.お前とて、指し間違えた手を待てと言っても聞かぬではないか。@k@v01/11500014.ならば@bあいこ.@<相子@>というものだろう」");

    assert_debug_snapshot!("ep1_kinzo1", snapshot);
}

#[test]
fn snapshot_ep1_kumasawa_nvl() {
    // tests novel mode, and, apparently, kinsoku shori-forbidden character on start of the line
    let snapshot = make_snapshot(MessageTextLayout::Justify, MessageboxType::Novel, "熊沢\u{3000}チヨ@r@v18/11800041_1.……おいたわしや、紗音さん、嘉音さん…。@k@r@v18/11800041_2.あの二人がいじめられる理由は何もないのです。@k@r@v18/11800041_3.…しかし、郷田さんに嫌われているのは紛れもない事実…。");

    assert_debug_snapshot!("ep1_kumasawa_nvl", snapshot);
}

#[test]
fn snapshot_ep2_beatrice_red_truth() {
    // tests red truth (instant text + color change), as well as section & sync commands, used to issue VM commands at certain points in the message
    let snapshot = make_snapshot(MessageTextLayout::Justify, MessageboxType::WitchSpace, "ベアトリーチェ@r@v27/20700938.「そうだ。@k@v27/20700939.だから妾はこれより、@v27/20700940.@|@y@c900.@[真実を語る時、赤を使うことにする@]@c.」");

    assert_debug_snapshot!("ep2_beatrice_red_truth", snapshot);
}

#[test]
fn snapshot_ep4_beatrice_blue_truth() {
    // tests 2 blue truths in one message
    let snapshot = make_snapshot(MessageTextLayout::Justify, MessageboxType::Novel, "ベアトリーチェ@r@v27/40700828.@|@y@c279.@[以上の復唱要求、並びに復唱拒否から、@]@k@v27/40700829.@|@y@[妾はそなたに対戦相手の資格がないことを宣言する。@]@c.");

    assert_debug_snapshot!("ep4_beatrice_blue_truth", snapshot);
}

#[test]
fn snapshot_ep1_novel_centered() {
    // test centered text in NVL mode
    let snapshot = make_snapshot(
        MessageTextLayout::Center,
        MessageboxType::Novel,
        "@r@a1666.ダカラ、@w666.現金ガスグニ大量ニ、@r@w666.喉カラ手ガ出ルホドニ欲シカッタ…！",
    );

    assert_debug_snapshot!("ep1_novel_centered", snapshot);
}

#[test]
fn snapshot_ep1_narrator1() {
    // this tests a non-novel message without a character name
    let snapshot = make_snapshot(
        MessageTextLayout::Justify,
        MessageboxType::Ushiromiya,
        "@r聴診器を外しながら、年輩の医師は溜め息を漏らす。",
    );

    assert_debug_snapshot!("ep1_narrator1", snapshot);
}

#[test]
fn snapshot_ep1_kinzo2() {
    // tests the correct implementation of multiline quoted text
    let snapshot = make_snapshot(
        MessageTextLayout::Justify,
        MessageboxType::Ushiromiya,
        "右代宮\u{3000}金蔵@r@v01/11500003.「忠告の気持ちだけはありがたくいただいておく。@k@v01/11500004.我が友よ。@k@v01/11500005.………源次。@k@v01/11500006.もう一杯頼む。@k@v01/11500007.心持ち薄めでな。@k@v01/11500008.南條の顔も立ててやれ」",
    );

    assert_debug_snapshot!("ep1_kinzo2", snapshot);
}

#[test]
fn snapshot_ep1_battler1() {
    // tests the correct handling of prohibition rules in combination with rubi text
    let snapshot = make_snapshot(
        MessageTextLayout::Justify,
        MessageboxType::Ushiromiya,
        "右代宮\u{3000}戦人@r@v10/10100014.「こんだけ@bた.@<立@>っ@bぱ.@<端@>がありゃー俺の発育は充分っすから～！@k\u{3000}@v10/10100015.むしろちょいと身長縮めた方が服が探しやすいくらい！」",
    );

    assert_debug_snapshot!("ep1_battler1", snapshot);
}

#[test]
fn snapshot_ep1_narrator2() {
    // tests the correctness of time reflowing when rubi and base texts have equal width
    let snapshot = make_snapshot(
        MessageTextLayout::Justify,
        MessageboxType::Ushiromiya,
        "@r@bねこぐるま.@<猫車@>のひとつしかない車輪が小石でも噛んだのでバランスを崩してしまったのだろう。",
    );

    assert_debug_snapshot!("ep1_narrator2", snapshot);
}

#[test]
fn snapshot_ep1_kinzo3() {
    // tests correctness of squish/justify selection when max_width = layout_width
    let snapshot = make_snapshot(
        MessageTextLayout::Justify,
        MessageboxType::Ushiromiya,
        "右代宮\u{3000}金蔵@r@v01/11500136.「留弗夫の間抜けは女遊びばかりッ！！@k\u{3000}@v01/11500137.楼座はどこの馬の骨ともわからん男の赤ん坊など生みおって！！@k\u{3000}@v01/11500138.朱志香は無能で無学だ！！@k\u{3000}@v01/11500139.譲治には男としての器がない！@k\u{3000}@v01/11500140.戦人は右代宮家の栄誉を自ら捨ておった愚か者だッ！！@k\u{3000}@v01/11500141.真里亞など見るのも汚らわしいッ！！」",
    );

    assert_debug_snapshot!("ep1_kinzo3", snapshot);
}
