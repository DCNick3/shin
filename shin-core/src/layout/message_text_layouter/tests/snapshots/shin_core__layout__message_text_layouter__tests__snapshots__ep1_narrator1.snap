---
source: shin-core/src/layout/message_text_layouter/tests/snapshots.rs
expression: snapshot
---
(
    [
        Char(
            Char {
                time: 0.0,
                line_index: 1,
                font: Regular,
                codepoint: '聴',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.000153,
                height: 50.0,
                position: Vec2(
                    0.0,
                    125.625,
                ),
                horizontal_scale: 0.50505203,
                scale: 0.5208333,
                color: UnormColor(
                    4294967295,
                ),
                fade: 0.2,
            },
        ),
        Char(
            Char {
                time: 0.025000075,
                line_index: 1,
                font: Regular,
                codepoint: '診',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.000153,
                height: 50.0,
                position: Vec2(
                    50.000153,
                    125.625,
                ),
                horizontal_scale: 0.50505203,
                scale: 0.5208333,
                color: UnormColor(
                    4294967295,
                ),
                fade: 0.2,
            },
        ),
        Char(
            Char {
                time: 0.05000015,
                line_index: 1,
                font: Regular,
                codepoint: '器',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.000153,
                height: 50.0,
                position: Vec2(
                    100.000305,
                    125.625,
                ),
                horizontal_scale: 0.50505203,
                scale: 0.5208333,
                color: UnormColor(
                    4294967295,
                ),
                fade: 0.2,
            },
        ),
        Char(
            Char {
                time: 0.07500023,
                line_index: 1,
                font: Regular,
                codepoint: 'を',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.000153,
                height: 50.0,
                position: Vec2(
                    150.00046,
                    125.625,
                ),
                horizontal_scale: 0.50505203,
                scale: 0.5208333,
                color: UnormColor(
                    4294967295,
                ),
                fade: 0.2,
            },
        ),
        Char(
            Char {
                time: 0.1000003,
                line_index: 1,
                font: Regular,
                codepoint: '外',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.000153,
                height: 50.0,
                position: Vec2(
                    200.00061,
                    125.625,
                ),
                horizontal_scale: 0.50505203,
                scale: 0.5208333,
                color: UnormColor(
                    4294967295,
                ),
                fade: 0.2,
            },
        ),
        Char(
            Char {
                time: 0.12500037,
                line_index: 1,
                font: Regular,
                codepoint: 'し',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.000153,
                height: 50.0,
                position: Vec2(
                    250.00076,
                    125.625,
                ),
                horizontal_scale: 0.50505203,
                scale: 0.5208333,
                color: UnormColor(
                    4294967295,
                ),
                fade: 0.2,
            },
        ),
        Char(
            Char {
                time: 0.15000045,
                line_index: 1,
                font: Regular,
                codepoint: 'な',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.000153,
                height: 50.0,
                position: Vec2(
                    300.00092,
                    125.625,
                ),
                horizontal_scale: 0.50505203,
                scale: 0.5208333,
                color: UnormColor(
                    4294967295,
                ),
                fade: 0.2,
            },
        ),
        Char(
            Char {
                time: 0.17500053,
                line_index: 1,
                font: Regular,
                codepoint: 'が',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.000153,
                height: 50.0,
                position: Vec2(
                    350.00107,
                    125.625,
                ),
                horizontal_scale: 0.50505203,
                scale: 0.5208333,
                color: UnormColor(
                    4294967295,
                ),
                fade: 0.2,
            },
        ),
        Char(
            Char {
                time: 0.20000061,
                line_index: 1,
                font: Regular,
                codepoint: 'ら',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: true,
                has_rubi: false,
                width: 50.000153,
                height: 50.0,
                position: Vec2(
                    400.00122,
                    125.625,
                ),
                horizontal_scale: 0.50505203,
                scale: 0.5208333,
                color: UnormColor(
                    4294967295,
                ),
                fade: 0.2,
            },
        ),
        Char(
            Char {
                time: 0.2250007,
                line_index: 1,
                font: Regular,
                codepoint: '、',
                is_rubi: false,
                cant_be_at_line_start: true,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.000153,
                height: 50.0,
                position: Vec2(
                    450.00137,
                    125.625,
                ),
                horizontal_scale: 0.50505203,
                scale: 0.5208333,
                color: UnormColor(
                    4294967295,
                ),
                fade: 0.2,
            },
        ),
        Char(
            Char {
                time: 0.3000008,
                line_index: 1,
                font: Regular,
                codepoint: '年',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.000153,
                height: 50.0,
                position: Vec2(
                    500.00153,
                    125.625,
                ),
                horizontal_scale: 0.50505203,
                scale: 0.5208333,
                color: UnormColor(
                    4294967295,
                ),
                fade: 0.2,
            },
        ),
        Char(
            Char {
                time: 0.32500085,
                line_index: 1,
                font: Regular,
                codepoint: '輩',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.000153,
                height: 50.0,
                position: Vec2(
                    550.0017,
                    125.625,
                ),
                horizontal_scale: 0.50505203,
                scale: 0.5208333,
                color: UnormColor(
                    4294967295,
                ),
                fade: 0.2,
            },
        ),
        Char(
            Char {
                time: 0.35000092,
                line_index: 1,
                font: Regular,
                codepoint: 'の',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.000153,
                height: 50.0,
                position: Vec2(
                    600.00183,
                    125.625,
                ),
                horizontal_scale: 0.50505203,
                scale: 0.5208333,
                color: UnormColor(
                    4294967295,
                ),
                fade: 0.2,
            },
        ),
        Char(
            Char {
                time: 0.37500098,
                line_index: 1,
                font: Regular,
                codepoint: '医',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.000153,
                height: 50.0,
                position: Vec2(
                    650.00195,
                    125.625,
                ),
                horizontal_scale: 0.50505203,
                scale: 0.5208333,
                color: UnormColor(
                    4294967295,
                ),
                fade: 0.2,
            },
        ),
        Char(
            Char {
                time: 0.40000105,
                line_index: 1,
                font: Regular,
                codepoint: '師',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.000153,
                height: 50.0,
                position: Vec2(
                    700.0021,
                    125.625,
                ),
                horizontal_scale: 0.50505203,
                scale: 0.5208333,
                color: UnormColor(
                    4294967295,
                ),
                fade: 0.2,
            },
        ),
        Char(
            Char {
                time: 0.4250011,
                line_index: 1,
                font: Regular,
                codepoint: 'は',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.000153,
                height: 50.0,
                position: Vec2(
                    750.0022,
                    125.625,
                ),
                horizontal_scale: 0.50505203,
                scale: 0.5208333,
                color: UnormColor(
                    4294967295,
                ),
                fade: 0.2,
            },
        ),
        Char(
            Char {
                time: 0.45000118,
                line_index: 1,
                font: Regular,
                codepoint: '溜',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.000153,
                height: 50.0,
                position: Vec2(
                    800.0023,
                    125.625,
                ),
                horizontal_scale: 0.50505203,
                scale: 0.5208333,
                color: UnormColor(
                    4294967295,
                ),
                fade: 0.2,
            },
        ),
        Char(
            Char {
                time: 0.47500125,
                line_index: 1,
                font: Regular,
                codepoint: 'め',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.000153,
                height: 50.0,
                position: Vec2(
                    850.00244,
                    125.625,
                ),
                horizontal_scale: 0.50505203,
                scale: 0.5208333,
                color: UnormColor(
                    4294967295,
                ),
                fade: 0.2,
            },
        ),
        Char(
            Char {
                time: 0.5000013,
                line_index: 1,
                font: Regular,
                codepoint: '息',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.000153,
                height: 50.0,
                position: Vec2(
                    900.00256,
                    125.625,
                ),
                horizontal_scale: 0.50505203,
                scale: 0.5208333,
                color: UnormColor(
                    4294967295,
                ),
                fade: 0.2,
            },
        ),
        Char(
            Char {
                time: 0.5250014,
                line_index: 1,
                font: Regular,
                codepoint: 'を',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.000153,
                height: 50.0,
                position: Vec2(
                    950.0027,
                    125.625,
                ),
                horizontal_scale: 0.50505203,
                scale: 0.5208333,
                color: UnormColor(
                    4294967295,
                ),
                fade: 0.2,
            },
        ),
        Char(
            Char {
                time: 0.5500015,
                line_index: 1,
                font: Regular,
                codepoint: '漏',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.000153,
                height: 50.0,
                position: Vec2(
                    1000.0028,
                    125.625,
                ),
                horizontal_scale: 0.50505203,
                scale: 0.5208333,
                color: UnormColor(
                    4294967295,
                ),
                fade: 0.2,
            },
        ),
        Char(
            Char {
                time: 0.5750016,
                line_index: 1,
                font: Regular,
                codepoint: 'ら',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.000153,
                height: 50.0,
                position: Vec2(
                    1050.0029,
                    125.625,
                ),
                horizontal_scale: 0.50505203,
                scale: 0.5208333,
                color: UnormColor(
                    4294967295,
                ),
                fade: 0.2,
            },
        ),
        Char(
            Char {
                time: 0.6000017,
                line_index: 1,
                font: Regular,
                codepoint: 'す',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: true,
                has_rubi: false,
                width: 50.000153,
                height: 50.0,
                position: Vec2(
                    1100.003,
                    125.625,
                ),
                horizontal_scale: 0.50505203,
                scale: 0.5208333,
                color: UnormColor(
                    4294967295,
                ),
                fade: 0.2,
            },
        ),
        Char(
            Char {
                time: 0.6250018,
                line_index: 1,
                font: Regular,
                codepoint: '。',
                is_rubi: false,
                cant_be_at_line_start: true,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.000153,
                height: 50.0,
                position: Vec2(
                    1150.0032,
                    125.625,
                ),
                horizontal_scale: 0.50505203,
                scale: 0.5208333,
                color: UnormColor(
                    4294967295,
                ),
                fade: 0.2,
            },
        ),
        Wait(
            Wait {
                time: 0.7500019,
                line_index: 1,
                is_last_wait: true,
                is_auto_click: false,
            },
        ),
    ],
    [
        LineInfo {
            width: 0.0,
            y_position: 0.0,
            line_height: 65.0,
            baseline_ascent: 60.625,
            rubi_height: 20.0,
        },
        LineInfo {
            width: 1200.0033,
            y_position: 65.0,
            line_height: 70.0,
            baseline_ascent: 60.625,
            rubi_height: 20.0,
        },
    ],
    Vec2(
        1200.0033,
        135.0,
    ),
)
